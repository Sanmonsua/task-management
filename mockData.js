export const category = {
    name : "Work",
    tasks : [
        {
            id : 1,
            name: 'Make changes to the new site design',
            date : '15 sept 2020',
            done : false,
        },
        {
            id : 2,
            name: 'Try new icon set',
            date : '11 sept 2020',
            done : true,
        },
    ],
    color : '#a362ea',
}