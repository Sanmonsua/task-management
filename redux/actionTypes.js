// action types
export const ADD_CATEGORY = 'ADD_CATEGORY'
export const ADD_TASK = 'ADD_TASK'