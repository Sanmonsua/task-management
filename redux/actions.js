import { ADD_CATEGORY, ADD_TASK, TOGGLE_TASK, EDIT_TASK, SELECT_CATEGORY, FETCH_CATEGORIES } from './actionTypes'
import {firebaseApp} from '../firebase'

let newCategoryId = 0 
let newTaskId = 0
// action creators
export const addCategory = newCategory => {
  newCategoryId ++
  return ({
      type: ADD_CATEGORY,
      payload: {
        ... newCategory, 
        id : `${newCategoryId}`,
        tasks : [],
      },
    }
  )
}

export const addTask = newTask => {
  newTaskId ++
  return ({
      type: ADD_TASK,
      payload: {
        ...newTask,
        id : `${newTaskId}`,
        done : false,
      },
    }
  )
}

export const editTask = task => {
  return ({
    type : EDIT_TASK,
    payload : task,
  })
}

export const toggleTask = task => ({
  type: TOGGLE_TASK,
  payload: task,
})

export const selectCategory = category =>({
  type: SELECT_CATEGORY,
  payload: category,
})

export const fetchCategories = () => dispatch => {
  
  firebaseApp.database().ref('/categories').on('value', snap =>{
    
    dispatch({
      type: FETCH_CATEGORIES,
      payload: {categories : snap.val()},
    })
  })
}