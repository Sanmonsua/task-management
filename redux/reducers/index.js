import { combineReducers } from 'redux'
import categories from './categories'
import user from './user'

export default combineReducers({ categories, user })