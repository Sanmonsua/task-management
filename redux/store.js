import { createStore } from "redux";
import rootReducer from "./reducers";
import { addCategory, addTask } from "./actions";

const store = createStore(rootReducer);

store.dispatch(
    addCategory({
        name: "Work",
        color: "#a362ea",
        tasks : [],
        id: 1,
    })
);

store.dispatch(
    addTask({
        id : 1,
        name: 'Make changes to the new site design',
        date : '15 sept 2020',
        done : true,
        categoryId : 1,
    })
)

export default store