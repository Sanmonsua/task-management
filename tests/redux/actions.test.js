import * as actions from '../../redux/actions'
import * as actionsTypes from '../../redux/actionTypes'

import { mockCategory, mockUser } from '../../mockData'

// SignIn action
describe('signIn returns actions', () => {
    it('handles user uid input', () => {
        expect(actions.signIn(mockUser)).toMatchSnapshot()
    })

    it('handles user uid empty', () => {
        expect(actions.signIn({ uid : '' })).toEqual({ 
            type: actionsTypes.SIGN_IN, 
            payload: { uid : '' } 
        })
    })
})

// AddCategory action
describe('addCategory returns actions', () => {
    
    it('handles new category', () => {
        expect(actions.addCategory(mockCategory)).toMatchSnapshot({
            payload : {
                id : expect.any(String)
            }
        })
    })
})

describe('editCategory returns action', () => {

    it('handles valid category', () => {
        expect(actions.editCategory(mockCategory)).toMatchSnapshot()
    })

})