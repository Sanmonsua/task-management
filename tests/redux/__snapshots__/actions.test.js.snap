// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categories actions handles addCategory action 1`] = `
Object {
  "payload": Object {
    "color": "#a362ea",
    "id": Any<String>,
    "name": "Work",
    "tasks": Array [],
  },
  "type": "ADD_CATEGORY",
}
`;

exports[`categories actions handles deleteCategory action 1`] = `
Object {
  "payload": Object {
    "color": "#a362ea",
    "name": "Work",
    "tasks": Array [
      Object {
        "date": "15 sept 2020",
        "done": false,
        "id": 1,
        "name": "Make changes to the new site design",
      },
      Object {
        "date": "11 sept 2020",
        "done": true,
        "id": 2,
        "name": "Try new icon set",
      },
    ],
  },
  "type": "DELETE_CATEGORY",
}
`;

exports[`categories actions handles editCategory action 1`] = `
Object {
  "payload": Object {
    "color": "#a362ea",
    "name": "Work",
    "tasks": Array [
      Object {
        "date": "15 sept 2020",
        "done": false,
        "id": 1,
        "name": "Make changes to the new site design",
      },
      Object {
        "date": "11 sept 2020",
        "done": true,
        "id": 2,
        "name": "Try new icon set",
      },
    ],
  },
  "type": "EDIT_CATEGORY",
}
`;

exports[`categories actions handles selectCategory action 1`] = `
Object {
  "payload": Object {
    "color": "#a362ea",
    "name": "Work",
    "tasks": Array [
      Object {
        "date": "15 sept 2020",
        "done": false,
        "id": 1,
        "name": "Make changes to the new site design",
      },
      Object {
        "date": "11 sept 2020",
        "done": true,
        "id": 2,
        "name": "Try new icon set",
      },
    ],
  },
  "type": "SELECT_CATEGORY",
}
`;

exports[`signIn returns actions handles user uid input 1`] = `
Object {
  "payload": Object {
    "uid": "admin123",
  },
  "type": "SIGN_IN",
}
`;

exports[`tasks actions handles addTask 1`] = `
Object {
  "payload": Object {
    "date": "1600662268405",
    "done": false,
    "id": Any<String>,
    "name": "Make changes to the new site design",
  },
  "type": "ADD_TASK",
}
`;

exports[`tasks actions handles deleteTask 1`] = `
Object {
  "payload": Object {
    "date": "1600662268405",
    "done": false,
    "id": "1600662268405",
    "name": "Make changes to the new site design",
  },
  "type": "DELETE_TASK",
}
`;

exports[`tasks actions handles editTask 1`] = `
Object {
  "payload": Object {
    "date": "1600662268405",
    "done": false,
    "id": "1600662268405",
    "name": "Make changes to the new site design",
  },
  "type": "EDIT_TASK",
}
`;

exports[`tasks actions handles toggleTask 1`] = `
Object {
  "payload": Object {
    "date": "1600662268405",
    "done": false,
    "id": "1600662268405",
    "name": "Make changes to the new site design",
  },
  "type": "TOGGLE_TASK",
}
`;
